src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java

```diff
@@ -0,0 +1,60 @@
+package net.minecraftforge.gui.capability;
+
+import io.netty.buffer.ByteBuf;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a gui provider- something that activates a gui for a player to use.
+ */
+public interface IGuiProvider
+{
+    /**
+     * Deserializes a buffer from the server into the provider instance.
+     *
+     * @param buffer
+     */
+    void deserialize(ByteBuf buffer);
+
+    /**
+     * Serializes data on the server to send along with the open gui packet.
+     * Use this to add any information you need to open the gui properly on the client.
+     *
+     * @param buffer
+     */
+    void serialize(ByteBuf buffer);
+
+    /**
+     * Gets the gui provider object from the world, player, etc.
+     * Used in the get*Element methods to pass a direct instance to them.
+     *
+     * @param player The player instance.
+     * @param world The world.
+     * @return
+     */
+    Object getOwner(EntityPlayer player, World world);
```

Comment:
The statefulness introduced here (`deserialize` and `getOwner` must be called after each other!) is not very nice imho.